libherokubuildpack/src/write.rs

```diff
@@ -0,0 +1,212 @@
+    fn drop(&mut self) {
+        // Drop implementations must not panic. We intentionally ignore the potential error here.
+        let _result = self.map_and_write_current_buffer();
+    }
+}
+
+impl<A: io::Write, B: io::Write> io::Write for TeeWrite<A, B> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner_a.write_all(buf)?;
+        self.inner_b.write_all(buf)?;
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner_a.flush()?;
+        self.inner_b.flush()
+    }
+}
+
+const NEWLINE_ASCII_BYTE: u8 = 0x0Au8;
+
+#[cfg(test)]
+mod test {
+    use super::tee;
+    use crate::write::line_mapped;
+
+    #[test]
+    fn test_tee_write() {
+        let mut a = vec![];
+        let mut b = vec![];
+
+        let mut input = "foo bar baz".as_bytes();
+        std::io::copy(&mut input, &mut tee(&mut a, &mut b)).unwrap();
+
+        assert_eq!(a, "foo bar baz".as_bytes());
+        assert_eq!(a, b);
+    }
+
+    #[test]
+    fn test_mapped_write() {
+        let mut output = vec![];
+
+        let mut input = "foo\nbar\nbaz".as_bytes();
+        std::io::copy(
+            &mut input,
+            &mut line_mapped(&mut output, |line| line.repeat(2)),
+        )
+        .unwrap();
+
+        assert_eq!(output, "foo\nfoo\nbar\nbar\nbazbaz".as_bytes());
+    }
+}
+
+/// Mapper functions for use with the [`mapped`] and [`line_mapped`] functions.
+pub mod mappers {
+    /// Adds a prefix.
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libherokubuildpack::write::line_mapped;
+    /// use libherokubuildpack::write::mappers::add_prefix;
+    /// use libherokubuildpack::command::CommandExt;
+    /// use std::process::Command;
+    ///
+    /// Command::new("date")
+    ///     .spawn_and_write_streams(
+    ///         line_mapped(
+    ///             std::io::stdout(),
+    ///             add_prefix("date stdout> "),
+    ///         ),
+    ///         std::io::stderr(),
+    ///     )
+    ///     .and_then(|mut child| child.wait())
+    ///     .unwrap();
+    /// ```
+    pub fn add_prefix<P: Into<Vec<u8>>>(prefix: P) -> impl Fn(Vec<u8>) -> Vec<u8> {
+        let prefix = prefix.into();
+
+        move |mut input| {
+            let mut result = prefix.clone();
+            result.append(&mut input);
+            result
+        }
+    }
+
+    /// Allows mapping the data as an UTF-8 string that was lossy converted from the data to be mapped.
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libherokubuildpack::write::line_mapped;
+    /// use libherokubuildpack::write::mappers::map_utf8_lossy;
+    /// use libherokubuildpack::command::CommandExt;
+    /// use std::process::Command;
+    ///
+    /// Command::new("date")
+    ///     .spawn_and_write_streams(
+    ///         line_mapped(
+    ///             std::io::stdout(),
+    ///             map_utf8_lossy(|string| string.replace("foo", "bar")),
+    ///         ),
```

Comment:
`output_and_write_streams` would contain the _original_ output. The manipulations via the `Write`s happens separately and _in addition_. This ensures `output_and_write_streams` is a drop-in replacement for `output`. `CommandExt` shouldn't really be a concern of this module, it only contains `Write` implementations that can be used wherever. I use `CommandExt` in the Rustdoc as an example because that is what they are designed for, but the implementation itself shouldn't be coupled to it.

If you want to capture the modified stream, a `&mut [u8]` is `Write`. So you can easily "pipe" it into a variable of that type and inspect it after the command finished.